pyudev: depend on udev instead of systemd

a-m-joseph opened the pull request:
Pyudev depends only on a libudev implementation, rather than on
    systemd-udevd specifically or on systemd generally.  Let's adjust the
    dependency list to make it more specific.  In particular, this allows
    pyudev to build and run with the overlay which substitutes eudev for
    udev.

    Pyudev is used by the auxiliary trackpad calibration tools included as
    part of libinput.  Because an enormous number of packages depend on
    libinput, this allows a very large number of packages to be used
    without systemd.  Note that libinput.so does not depend on pyudev or
    systemd -- only the trackpad calibration utilities bundled with
    libudev use pyudev.


###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

a-m-joseph commented:
Latest force-push adds a commit message, no other changes.

a-m-joseph commented:
Ping...

a-m-joseph commented:
Ping...

jtojnar commented:
This does not make much sense to me. `libudev` is just a part of systemd these days, so the option will just break the package.

a-m-joseph commented:
> so the option will just break the package.

Not true -- it compiles just fine with `eudev`.

jtojnar commented:
Sure, but that is not what this patch does. If you do not hardcode the `libudev` path, the package will fail at runtime, unless you set `LD_LIBRARY_PATH` environment variable appropriately.

a-m-joseph commented:
You are correct.  I will revise this.

The big issue here is that libinput is bundled with a few python utilities for trackpad calibration, which use pyudev.  Almost everything depends on libinput-lib (but not these trackpad calibration utilities), so almost everything ends up dragging in systemd via libinput.

I am marking this as a draft in order to change the dependency from `systemd` to `udev`.  If that doesn't work I'll close this PR and open a different one to make the trackpad calibration utilities optional.

Commit:
pyudev: depend on udev instead of systemd

Pyudev depends only on a libudev implementation, rather than on
systemd-udevd specifically or on systemd generally.  Let's adjust the
dependency list to make it more specific.  In particular, this allows
pyudev to build and run with the overlay which substitutes eudev for
udev.

Pyudev is used by the auxiliary trackpad calibration tools included as
part of libinput.  Because an enormous number of packages depend on
libinput, this allows a very large number of packages to be used
without systemd.  Note that libinput.so does not depend on pyudev or
systemd -- only the trackpad calibration utilities bundled with
libudev use pyudev.

a-m-joseph commented:
Updated and re-titled.

I have verified that this evaluates, builds, and that the trackpad utilities no longer crash as they did when the expression simply omitted any source for `libudev.so`.  However, the machine I am sitting at right now has no trackpad, so I get the expected

```
$ /nix/store/zqrsamnlna369gm795vwdcwsba9s4kmp-libinput-1.19.3-bin/libexec/libinput/libinput-measure-fuzz
Unable to find a touch device.
```

I will double-check this using a laptop (which has a libinput-compatible trackpad) tomorrow.

jtojnar commented:
Thanks.
